mono/metadata/gc.c

```diff
@@ -381,8 +381,10 @@ mono_object_register_finalizer (MonoObject *obj)
 gboolean
 mono_domain_finalize (MonoDomain *domain, guint32 timeout) 
 {
-	DomainFinalizationReq *req;
```

Comment:
So the problem here is that the thread we are waiting for might take longer than the timeout, which means that
we cannot wait for it to terminate, sometimes the other thread needs to clean the data. So it would be better
to put a refcount field into DomainFinalizationReq, initialize it to 2, and atomically decrease it when one of the threads is done with it, then free it (from both threads) when the refcount gets to 0.